spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -869,26 +882,45 @@ def _retrieve_synaptic_block(
         block = None
         if max_row_length > 0 and synaptic_block_offset is not None:
 
+            # if exploiting the extra monitor cores, need to set the machine
+            # for data extraction mode
+            if using_extra_monitor_cores and handle_time_out_configuration:
+                data_receiver.set_cores_for_data_extraction(
+                    transceiver, extra_monitor_cores_for_router_timeout,
+                    placements)
+
             if not is_single:
                 # calculate the synaptic block size in bytes
                 synaptic_block_size = self._synapse_io.get_block_n_bytes(
                     max_row_length, n_rows)
 
                 # read in the synaptic block
-                block = transceiver.read_memory(
-                    placement.x, placement.y,
-                    indirect_synapses_address + synaptic_block_offset,
-                    synaptic_block_size)
+                if using_extra_monitor_cores:
+                    block = data_receiver.get_data(
+                        transceiver, sender_extra_monitor_core_placement,
+                        indirect_synapses_address + synaptic_block_offset,
+                        synaptic_block_size)
+                else:
+                    block = transceiver.read_memory(
+                        placement.x, placement.y,
+                        indirect_synapses_address + synaptic_block_offset,
+                        synaptic_block_size)
 
             else:
                 # The data is one per row
                 synaptic_block_size = n_rows * 4
 
-                # read in the synaptic row data
-                single_block = numpy.asarray(transceiver.read_memory(
-                    placement.x, placement.y,
-                    direct_synapses_address + synaptic_block_offset,
-                    synaptic_block_size), dtype="uint8").view("uint32")
+                if using_extra_monitor_cores:
+                    # read in the synaptic row data
+                    single_block = numpy.asarray(data_receiver.get_data(
+                        transceiver, sender_extra_monitor_core_placement,
+                        direct_synapses_address + synaptic_block_offset,
+                        synaptic_block_size), dtype="uint8").view("uint32")
+                else:
+                    single_block = numpy.asarray(transceiver.read_memory(
+                        placement.x, placement.y,
+                        direct_synapses_address + synaptic_block_offset,
+                        synaptic_block_size), dtype="uint8").view("uint32")
```

Comment:
I see this pattern in multiple places in this method. Refactoring opportunity?

Reply:
annoyingly you gain nothing doing the refactor. its actually more lines. as your refactoring one line in the first place. most of it is parameters.